File: cmd/help.py
import discord
from discord.ext import commands
from discord import app_commands

from utils.embeds import make_embed
from utils.check_perms import is_bot_admin
from db.db_helpers.commands import is_command_disabled


class Help(commands.Cog):

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @app_commands.command(name="help",
                          description="Show available bot commands")
    async def help(self, interaction: discord.Interaction):
        guild = interaction.guild
        is_admin = is_bot_admin(interaction)

        command_groups: dict[str, list[str]] = {}

        for command in self.bot.tree.get_commands():
            if guild and is_command_disabled(guild.id, command.name):
                if not is_admin:
                    continue

            # Admin-only heuristic (commands that manage server/bot)
            admin_only = command.name.startswith(
                ("adminrole", "sticky", "command_", "addrole", "removerole"))

            if admin_only and not is_admin:
                continue

            group = command.parent.name if command.parent else "General" # type: ignore

            entry = f"`/{command.name}` – {command.description or 'No description'}" # type: ignore
            command_groups.setdefault(group, []).append(entry)

        fields = []
        for group, cmds in sorted(command_groups.items()):
            fields.append((group, "\n".join(sorted(cmds)), False))

        embed = make_embed(
            title="Help",
            description="Available slash commands",
            level="INFO",
            fields=fields,
            footer="Use slash commands to interact with the bot")

        await interaction.response.send_message(embed=embed, ephemeral=True)


async def setup(bot: commands.Bot):
    await bot.add_cog(Help(bot))

File: src/commands/utility/help.js
import { PermissionFlagsBits, SlashCommandBuilder } from "discord.js";
import { CONFIG } from "../../config.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { EMOJIS } from "../../core/emojis.js";
import { getRestrictedCommands } from "../../db/helpers/channelCommandRestrict.js";

const slashBuilder = new SlashCommandBuilder()
  .setName("help")
  .setDescription("View bot commands and usage instructions")
  .addStringOption((opt) => opt.setName("command").setDescription("Specific command to inspect").setRequired(false));

export default createCommand({
  name: "help",
  description: "View bot commands and usage instructions",
  category: "Utility",
  slashBuilder,

  async execute(ctx) {
    const { client, guild, channel, member } = ctx;
    const query = ctx.options.command || ctx.options.primary;

    // A. Specific Command Details
    if (query) {
      const cmdName = query.toLowerCase().replace(/^\//, "");
      const resolvedName = client.aliases.get(cmdName) || cmdName;
      const cmd = client.commands.get(resolvedName);

      if (!cmd) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "Command Not Found",
              description: `${EMOJIS.get("fail") || "❌"} No command found matching \`${query}\`.`,
              level: "ERROR",
            }),
          ],
          ephemeral: true,
        });
      }

      const desc =
        `**Description:** ${cmd.description}\n` +
        `**Category:** ${cmd.category}\n` +
        `**Aliases:** ${cmd.aliases.length > 0 ? cmd.aliases.map((a) => `\`${a}\``).join(", ") : "*None*"}\n` +
        `**Authority:** ${cmd.adminOnly ? "`Bot Admin`" : cmd.configOnly ? "`Manage Server`" : cmd.modOnly ? "`Moderator`" : "`Everyone`"}`;

      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: `Help • /${cmd.name}`,
            description: desc,
            level: "INFO",
          }),
        ],
      });
    }

    // B. Full Directory Overview
    let restrictedSet = new Set();
    const isAdmin = member?.permissions?.has(PermissionFlagsBits.Administrator);
    if (guild && channel && !isAdmin) {
      const restricted = await getRestrictedCommands(guild.id, channel.id);
      restrictedSet = new Set(restricted.map((c) => c.toLowerCase()));
    }

    const categories = new Map();
    for (const cmd of client.commands.values()) {
      if (restrictedSet.has(cmd.name.toLowerCase())) continue;

      const cat = cmd.category || "Utility";
      if (!categories.has(cat)) categories.set(cat, []);
      categories.get(cat).push(cmd.name);
    }

    const prefix = CONFIG.PREFIX || "ts";
    const botName = CONFIG.BOT_NAME || client.user?.username || "Digital Vigil";

    const embed = makeEmbed({
      title: `${botName} • Commands Directory`,
      description:
        `${EMOJIS.get("announcement") || "📌"} Use \`${prefix}help [command]\` or \`/help [command]\` for details.\n\n` +
        `*Slash commands (\`/\`) and prefix commands (\`${prefix}\`) are both supported.*`,
      level: "INFO",
      footer: `Total Available Commands: ${[...categories.values()].flat().length}`,
    });

    for (const [cat, cmds] of categories.entries()) {
      const formatted = cmds.sort().map((c) => `\`${c}\``).join(" ");
      embed.addFields({
        name: `**${cat} (${cmds.length})**`,
        value: formatted || "*No commands*",
        inline: false,
      });
    }

    return await ctx.reply({ embeds: [embed] });
  },
});
